pandas/plotting/_matplotlib/core.py

```diff
@@ -240,18 +246,28 @@ def _iter_data(self, data=None, keep_index=False, fillna=None):
         if fillna is not None:
             data = data.fillna(fillna)
 
-        for col, values in data.items():
-            if keep_index is True:
-                yield col, values
-            else:
-                yield col, values.values
+        iter_data = data
+        if self.by is not None:
+            # select sub-columns based on the value of first level of MI
+            cols = data.columns.levels[0]
+            iter_data = {
+                col: data.loc[:, data.columns.get_level_values(0) == col]
+                for col in cols
+            }
+
+        for col, val in iter_data.items():
+            if not keep_index:
+                val = val.values
+            yield col, val
 
     @property
     def nseries(self):
         if self.data.ndim == 1:
             return 1
-        else:
+        elif self.by is None:
             return self.data.shape[1]
+        else:
+            return len(self.data.columns.levels[0])
```

Comment:
we don't want to reach *into* the internals this much, IOW don't use levels, try using groupby instead

Reply:
nice! changed!